File: packages/silo-client-java/src/main/java/in/org/quicko/silo/client/transport/PagePayload.java
package in.org.quicko.silo.client.transport;

import com.fasterxml.jackson.databind.JsonNode;
import java.util.ArrayList;
import java.util.List;

/**
 * Reads a list response body. The wire uses {@code data} for entries and search
 * and {@code items} for media, collections and projects — this is the only place
 * in the client that knows both names.
 *
 * <p>{@code limit} and {@code offset} are null when the route echoed no window,
 * which the caller then falls back on the requested one for.
 */
public record PagePayload(List<JsonNode> rows, int total, Integer limit, Integer offset) {

  public static PagePayload read(JsonNode body) {
    JsonNode array = body.has("data") && body.get("data").isArray() ? body.get("data")
        : body.has("items") && body.get("items").isArray() ? body.get("items") : null;

    List<JsonNode> rows = new ArrayList<>();
    if (array != null) array.forEach(rows::add);

    return new PagePayload(
        List.copyOf(rows),
        body.path("total").isNumber() ? body.get("total").asInt() : rows.size(),
        body.path("limit").isNumber() ? body.get("limit").asInt() : null,
        body.path("offset").isNumber() ? body.get("offset").asInt() : null);
  }
}

File: packages/silo-client-dart/lib/src/transport/page_payload.dart
import 'json_values.dart';

/// A list body. Entries and search use `data`; everything else uses `items`.
/// [limit] and [offset] are null when the route echoed no window.
final class PagePayload {
  factory PagePayload.read(Map<String, Object?> body) {
    final rows = JsonValues.objects(body['data'] ?? body['items']);
    return PagePayload._(
      rows,
      body['total'] is int ? body['total'] as int : rows.length,
      body['limit'] as int?,
      body['offset'] as int?,
    );
  }

  const PagePayload._(this.rows, this.total, this.limit, this.offset);

  final List<Map<String, Object?>> rows;
  final int total;
  final int? limit;
  final int? offset;
}
